src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVM.cpp

```diff
@@ -106,128 +106,128 @@ void SinglePhaseFVM< BASE >::setupSystem( DomainPartition & domain,
+                                                     localRhs,
+                                                     subRegion,
+                                                     fluid,
+                                                     solid,
+                                                     subRegionFlowResidualNorm,
+                                                     subRegionFlowResidualNormalizer );
+        subRegionResidualNorm[0] = subRegionFlowResidualNorm[0];
+        subRegionResidualNormalizer[0] = subRegionFlowResidualNormalizer[0];
       }
-    } );
-
-    // compute global residual norm
-
-    real64 globalFlowResidualNorm[3] = {0, 0, 0};
-
-    MpiWrapper::allReduce( localFlowResidualNorm,
-                           globalFlowResidualNorm,
-                           3,
-                           MPI_SUM,
-                           MPI_COMM_GEOSX );
-
-    residualFlowAllMeshes += sqrt( globalFlowResidualNorm[0] ) / ( ( globalFlowResidualNorm[1] + m_fluxEstimate ) / (globalFlowResidualNorm[2]+1) );
-
-    if( m_isThermal )
-    {
-      real64 globalEnergyResidualNorm[3] = {0, 0, 0};
 
-      MpiWrapper::allReduce( localEnergyResidualNorm,
-                             globalEnergyResidualNorm,
-                             3,
-                             MPI_SUM,
-                             MPI_COMM_GEOSX );
+      // step 2: first reduction across meshBodies/regions/subRegions
 
-      residualEnergyAllMeshes += sqrt( globalEnergyResidualNorm[0] ) / ( globalEnergyResidualNorm[1] / (globalEnergyResidualNorm[2]+1));
-    }
-
-    numMeshTargets++;
+      for( integer i = 0; i < numNorm; ++i )
+      {
+        if( BASE::m_normType == solverBaseKernels::NormType::Linf )
+        {
+          if( subRegionResidualNorm[i] > localResidualNorm[i] )
+          {
+            localResidualNorm[i] = subRegionResidualNorm[i];
+          }
+        }
+        else
+        {
+          localResidualNorm[i] += subRegionResidualNorm[i];
+          localResidualNormalizer[i] += subRegionResidualNormalizer[i];
+        }
+      }
+    } );
   } );
 
+  // step 3: second reduction across MPI ranks
+
+  real64 residualNorm = 0.0;
   if( m_isThermal )
   {
-    real64 const flowResidual = residualFlowAllMeshes / numMeshTargets;
-    real64 const energyResidual = residualEnergyAllMeshes / numMeshTargets;
-
-    residual = std::sqrt( flowResidual*flowResidual + energyResidual*energyResidual );
+    real64 const flowResidualNorm = ( BASE::m_normType == solverBaseKernels::NormType::Linf )
+      ? MpiWrapper::max( localResidualNorm[0] )
+      : sqrt( MpiWrapper::sum( localResidualNorm[0] ) ) / MpiWrapper::sum( localResidualNormalizer[0] );
+    real64 const energyResidualNorm = ( BASE::m_normType == solverBaseKernels::NormType::Linf )
+      ? MpiWrapper::max( localResidualNorm[1] )
+      : sqrt( MpiWrapper::sum( localResidualNorm[1] ) ) / MpiWrapper::sum( localResidualNormalizer[1] );
```

Comment:
That could be nice to get OOP here too.

Reply:
Some here, I used helpers:
https://github.com/GEOSX/GEOSX/blob/3eada543329ce0bd54db2f4d7adc0f42886f5cb9/src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVM.cpp#L203-L210
so not really OOP, but slightly better.